File: docs/breez-sdk/snippets/go/lightning_address.go
package example

import (
	"errors"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func ConfigLightningAddress() *breez_sdk_spark.Config {
	// ANCHOR: config-lightning-address
	lnurlDomain := "yourdomain.com"
	apiKey := "your-api-key"
	config := breez_sdk_spark.DefaultConfig(breez_sdk_spark.NetworkMainnet)
	config.ApiKey = &apiKey
	config.LnurlDomain = &lnurlDomain
	// ANCHOR_END: config-lightning-address
	return &config
}

func CheckLightningAddressAvailability(client *breez_sdk_spark.BreezClient) (bool, error) {
	username := "myusername"

	// ANCHOR: check-lightning-address
	request := breez_sdk_spark.CheckLightningAddressRequest{
		Username: username,
	}

	isAvailable, err := client.CheckLightningAddressAvailable(request)
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return false, err
	}
	// ANCHOR_END: check-lightning-address

	return isAvailable, nil
}

func RegisterLightningAddress(client *breez_sdk_spark.BreezClient) (*breez_sdk_spark.LightningAddressInfo, error) {
	username := "myusername"
	description := "My Lightning Address"

	// ANCHOR: register-lightning-address
	request := breez_sdk_spark.RegisterLightningAddressRequest{
		Username:    username,
		Description: &description,
	}

	addressInfo, err := client.RegisterLightningAddress(request)
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	_ = addressInfo.LightningAddress
	_ = addressInfo.Lnurl.Url
	_ = addressInfo.Lnurl.Bech32
	// ANCHOR_END: register-lightning-address

	return &addressInfo, nil
}

func GetLightningAddress(client *breez_sdk_spark.BreezClient) (*breez_sdk_spark.LightningAddressInfo, error) {
	// ANCHOR: get-lightning-address
	addressInfoOpt, err := client.GetLightningAddress()
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	if addressInfoOpt != nil {
		_ = addressInfoOpt.LightningAddress
		_ = addressInfoOpt.Username
		_ = addressInfoOpt.Description
		_ = addressInfoOpt.Lnurl.Url
		_ = addressInfoOpt.Lnurl.Bech32
	}
	// ANCHOR_END: get-lightning-address

	return addressInfoOpt, nil
}

func DeleteLightningAddress(client *breez_sdk_spark.BreezClient) error {
	// ANCHOR: delete-lightning-address
	err := client.DeleteLightningAddress()
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}
	// ANCHOR_END: delete-lightning-address

	return nil
}

func AccessSenderComment(client *breez_sdk_spark.BreezClient) error {
	paymentID := "<payment id>"
	response, err := client.GetPayment(breez_sdk_spark.GetPaymentRequest{
		PaymentId: paymentID,
	})
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}
	payment := response.Payment

	// ANCHOR: access-sender-comment
	// Check if this is a lightning payment with LNURL receive metadata
	if lightningDetails, ok := (*payment.Details).(breez_sdk_spark.PaymentDetailsLightning); ok {
		metadata := lightningDetails.LnurlReceiveMetadata

		// Access the sender comment if present
		if metadata != nil && metadata.SenderComment != nil {
			println("Sender comment:", *metadata.SenderComment)
		}
	}
	// ANCHOR_END: access-sender-comment
	return nil
}

func AccessNostrZap(client *breez_sdk_spark.BreezClient) error {
	paymentID := "<payment id>"
	response, err := client.GetPayment(breez_sdk_spark.GetPaymentRequest{
		PaymentId: paymentID,
	})
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}
	payment := response.Payment

	// ANCHOR: access-nostr-zap
	// Check if this is a lightning payment with LNURL receive metadata
	if lightningDetails, ok := (*payment.Details).(breez_sdk_spark.PaymentDetailsLightning); ok {
		metadata := lightningDetails.LnurlReceiveMetadata

		if metadata != nil {
			// Access the Nostr zap request if present
			if metadata.NostrZapRequest != nil {
				// The NostrZapRequest is a JSON string containing the Nostr event (kind 9734)
				println("Nostr zap request:", *metadata.NostrZapRequest)
			}

			// Access the Nostr zap receipt if present
			if metadata.NostrZapReceipt != nil {
				// The NostrZapReceipt is a JSON string containing the Nostr event (kind 9735)
				println("Nostr zap receipt:", *metadata.NostrZapReceipt)
			}
		}
	}
	// ANCHOR_END: access-nostr-zap
	return nil
}

File: docs/breez-sdk/snippets/rust/src/lightning_address.rs
use breez_sdk_spark::{
    default_config, AcceptLightningAddressTransferRequest, BreezSdk, CheckLightningAddressRequest,
    Config, GetPaymentRequest, LightningAddressTransfer, Network, PaymentDetails,
    RegisterLightningAddressRequest,
};

pub fn configure_lightning_address() -> Config {
    // ANCHOR: config-lightning-address
    let mut config = default_config(Network::Mainnet);
    config.api_key = Some("your-api-key".to_string());
    config.lnurl_domain = Some("yourdomain.com".to_string());
    // ANCHOR_END: config-lightning-address
    config
}

pub async fn check_lightning_address_availability(sdk: &BreezSdk) -> anyhow::Result<bool> {
    // Define the username
    let username = "a username".to_string();

    // ANCHOR: check-lightning-address
    let request = CheckLightningAddressRequest { username };

    let is_available = sdk.check_lightning_address_available(request).await?;
    // ANCHOR_END: check-lightning-address
    Ok(is_available)
}

pub async fn register_lightning_address(sdk: &BreezSdk) -> anyhow::Result<(String, String, String)> {
    // Define the parameters
    let username = "a username".to_string();
    let description = Some("Lightning address description".to_string());

    // ANCHOR: register-lightning-address
    let request = RegisterLightningAddressRequest {
        username,
        description,
        transfer: None,
    };

    let address_info = sdk.register_lightning_address(request).await?;
    let lightning_address = address_info.lightning_address;
    let lnurl_url = address_info.lnurl.url;
    let lnurl_bech32 = address_info.lnurl.bech32;
    // ANCHOR_END: register-lightning-address
    Ok((lightning_address, lnurl_url, lnurl_bech32))
}

// Step 1 — run on the *current owner's* wallet. Produces the authorization
// the new owner needs to take over the username in a single atomic call.
pub async fn sign_lightning_address_transfer(
    current_owner_sdk: &BreezSdk,
    transferee_pubkey: &str,
) -> anyhow::Result<LightningAddressTransfer> {
    // ANCHOR: sign-lightning-address-transfer
    let transfer = current_owner_sdk
        .accept_lightning_address_transfer(AcceptLightningAddressTransferRequest {
            transferee_pubkey: transferee_pubkey.to_string(),
        })
        .await?;
    // ANCHOR_END: sign-lightning-address-transfer
    Ok(transfer)
}

// Step 2 — run on the *new owner's* wallet with the authorization received
// from the current owner (e.g. via QR code or deep link).
pub async fn register_lightning_address_via_transfer(
    new_owner_sdk: &BreezSdk,
    transfer: LightningAddressTransfer,
) -> anyhow::Result<(String, String, String)> {
    let username = "a username".to_string();
    let description = Some("Lightning address description".to_string());

    // ANCHOR: register-lightning-address-transfer
    let request = RegisterLightningAddressRequest {
        username,
        description,
        transfer: Some(transfer),
    };

    let address_info = new_owner_sdk.register_lightning_address(request).await?;
    let lightning_address = address_info.lightning_address;
    let lnurl_url = address_info.lnurl.url;
    let lnurl_bech32 = address_info.lnurl.bech32;
    // ANCHOR_END: register-lightning-address-transfer
    Ok((lightning_address, lnurl_url, lnurl_bech32))
}

pub async fn delete_lightning_address(sdk: &BreezSdk) -> anyhow::Result<()> {
    // ANCHOR: delete-lightning-address
    sdk.delete_lightning_address().await?;
    // ANCHOR_END: delete-lightning-address
    Ok(())
}

pub async fn get_lightning_address(sdk: &BreezSdk) -> anyhow::Result<()> {
    // ANCHOR: get-lightning-address
    let address_info_opt = sdk.get_lightning_address().await?;

    if let Some(info) = address_info_opt {
        let lightning_address = &info.lightning_address;
        let username = &info.username;
        let description = &info.description;
        let lnurl_url = &info.lnurl.url;
        let lnurl_bech32 = &info.lnurl.bech32;
    }
    // ANCHOR_END: get-lightning-address
    Ok(())
}

pub async fn access_sender_comment(sdk: &BreezSdk) -> anyhow::Result<()> {
    let payment_id = "<payment id>".to_string();
    let response = sdk.get_payment(GetPaymentRequest { payment_id }).await?;
    let payment = response.payment;

    // ANCHOR: access-sender-comment
    // Check if this is a lightning payment with LNURL receive metadata
    if let Some(PaymentDetails::Lightning {
        lnurl_receive_metadata: Some(metadata),
        ..
    }) = payment.details
    {
        // Access the sender comment if present
        if let Some(comment) = metadata.sender_comment {
            println!("Sender comment: {}", comment);
        }
    }
    // ANCHOR_END: access-sender-comment
    Ok(())
}

pub async fn access_nostr_zap(sdk: &BreezSdk) -> anyhow::Result<()> {
    let payment_id = "<payment id>".to_string();
    let response = sdk.get_payment(GetPaymentRequest { payment_id }).await?;
    let payment = response.payment;

    // ANCHOR: access-nostr-zap
    // Check if this is a lightning payment with LNURL receive metadata
    if let Some(PaymentDetails::Lightning {
        lnurl_receive_metadata: Some(metadata),
        ..
    }) = payment.details
    {
        // Access the Nostr zap request if present
        if let Some(zap_request) = metadata.nostr_zap_request {
            // The zap_request is a JSON string containing the Nostr event (kind 9734)
            println!("Nostr zap request: {}", zap_request);
        }

        // Access the Nostr zap receipt if present
        if let Some(zap_receipt) = metadata.nostr_zap_receipt {
            // The zap_receipt is a JSON string containing the Nostr event (kind 9735)
            println!("Nostr zap receipt: {}", zap_receipt);
        }
    }
    // ANCHOR_END: access-nostr-zap
    Ok(())
}
